clarification in multiEquatError()

weirichs opened the pull request:
The old function lacked a check if the testlets are constant. In addition, the comparison of the original and chained linking errors was misleading. Both should now be corrected.

weirichs committed:
clarification in multiEquatError()

The old function lacked a check if the testlets are constant. In addition, the comparison of the original and chained linking errors was misleading. Both should now be corrected.

sachseka commented on man/multiEquatError.rd:
```diff
@@ -48,12 +49,8 @@ Logical: print linking informations to console?
+A list of data.frames with linking errors, according to the number of domains. The output is
+intended to be used as argument of the replaceLinkingError function.
 }
 \author{
```
Maybe add Sebastian Weirich hier?

sachseka replied:
This change in authorship is mandatory because this is much more your code than mine.

sachseka reviewed (commented):
Without further documentation, it's unclear what has changed in the calculations. E.g., the results for Example 4 differ significantly:

Old version:
trend31  trend3221  le31   le3221  
1  0.1488454  0.2160025  0.07918052  0.1804314  

New version:
trend31  trend3221  le31   le3221  
1  0.1187884  0.2628832  0.05517095  0.148177  

Could you clarify why these differences occur? Adding comments to your code and writing tests would help ensure correctness and improve maintainability.

Would you like me to investigate the calculations myself, or do you have insights on the expected behavior??

weirichs commented:
> Old version: trend31 trend3221 le31 le3221 1 0.1488454 0.2160025 0.07918052 0.1804314
> 
> New version: trend31 trend3221 le31 le3221 1 0.1187884 0.2628832 0.05517095 0.148177

I'm sorry, this was a typo in the code. This should now be corrected. 

> Adding comments to your code and writing tests would help ensure correctness and improve maintainability.

I agree. Due to time constraints, it has not yet been possible to add meaningful tests or documentation. Basically, you would need an external criterion against which to compare it, otherwise you only ever look at consistency with previous versions, which could also be incorrect. In the long term, however, I would like to add documentation.

weirichs committed:
typo correction in multiEquatError()

weirichs commented:
P.S.: There is some weird behavior of gitKraken that causes large parts of the code to be displayed as modified, even though they have remained unchanged. (Nothing has changed in get.r) I hope this does not lead to you having to do a code review for all these sections that have remained identical. 

![Unbenannt](https://github.com/user-attachments/assets/89699bb6-fec9-420c-9e37-bcaeea378bbd)

sachseka commented:
> P.S.: There is some weird behavior of gitKraken that causes large parts of the code to be displayed as modified, even though they have remained unchanged. (Nothing has changed in get.r) I hope this does not lead to you having to do a code review for all these sections that have remained identical.

Yes, unfortunately, that is the case. So I'm not sure if I can meaningfully complete the code review because there are huge sections marked as changed that apparently haven't changed, but I will do my best

sachseka commented on R/multiEquatError.r:
```diff
@@ -1,206 +1,171 @@
             if (setdiff(colnames(testletStr), "item") != "unit") {
-                cat(paste0("The 'unit' column in 'testletStr' seemed to be called '",
-                           setdiff(colnames(testletStr), "item"),"'. Rename this column into 'unit'.\n"))
+                cat(paste0("The 'unit' column in 'testletStr' seemed to be called '",setdiff(colnames(testletStr), "item"),"'. Rename this column into 'unit'.\n"))
```
It is unclear whether this is a request to the user or if everything went well and this is done automatically (which is the case I think).

weirichs replied:
Yes, it's done automatically. I will rephrase the message.

sachseka commented on R/multiEquatError.r:
```diff
@@ -1,206 +1,171 @@
-                message(paste0("Following items without testlet definition: '",
-                               paste(setdiff(allI,testletStr[,"item"]) , collapse="', '"),
-                               "'.\n    Linking errors will be compute without considering testlets."))
+                message(paste0("Following items without testlet definition: '",paste(setdiff(allI,testletStr[,"item"]) , collapse="', '"), "'.\n    Linking errors will be compute without considering testlets."))
```
ok, this message to the user works

sachseka replied:
but I don't see the linkerrors change conditional on if testletstr is used here or not

weirichs replied:
link errors only will change if there is a _positive_ covariance between the DIF of measurement occation 1 vs. 2 and 2 vs. 3. Otherwise, linking errors should be identical.

sachseka replied:
this is not concerning the difference between 31 and 3221 but rather between JK1 and analytical linking error! The function does not switch to JK1 when it is supposed to do.

weirichs replied:
I think, it does. If testlets are specified, they are checked via `checkInputConsistency()`, and if the object returned is not NULL, the function switches into the `equa.rasch.jk` section. However, I think there is a bug in the original version of the package when applying the jackknife procedure: 

![Unbenannt](https://github.com/user-attachments/assets/e85723f0-8ef8-487a-bd05-c38fe5128eb1)

The highlighted passage should be `for(un in unique(e1223$unit))`, right?

sachseka replied:
I cannot check this right now, please revisit my code above inbetween. there it becomes clear, that JK-linkerrors are not used.

sachseka commented on R/multiEquatError.r:
```diff
@@ -1,206 +1,171 @@
+# hilfsfunktion (nicht auf NAMESPACE exportieren)
+tripleEquatError <- function(e1, e2, e3, dependentDIF, testletStr, difBound, dimName, verbose) {
+### checks
+  chk1 <- checkInputConsistency(e1=e1,e2=e2,e3=e3,testletStr=testletStr)    ### Die Rueckgabe von 'checkInputConsistency' ist NUR die Testletstruktur.
```
I think, it is somewhat unhealthy that completely different check objects are names chk1 (see below)

sachseka commented on R/multiEquatError.r:
```diff
@@ -1,206 +1,171 @@
 prepareAndCheckEatModelObject <- function ( liste, difBound, verbose ) {
+    ### all models unidim?
        uni  <- sapply(liste, FUN = function ( x ) { "correlation" %in% x[,"par"] })
        chk1 <- which(uni == TRUE)
```
this is the other chk1. please rename

sachseka commented on R/multiEquatError.r:
```diff
@@ -1,206 +1,171 @@
-       chk1<- lapply(comp, FUN = function ( com ){ if (length(intersect(it[[com[1]]], it[[com[2]]])) < 2) {
-         stop(paste0(length(intersect(it[[com[1]]], it[[com[2]]])) ," common items between measurement occasions ",
-                     com[1], " and ", com[2], "."))} })
+       chk1<- lapply(comp, FUN = function ( com ){ if ( length(intersect(it[[com[1]]], it[[com[2]]])) < 2) { stop(paste0(length(intersect(it[[com[1]]], it[[com[2]]])) ," common items between measurement occasions ",com[1], " and ", com[2], "."))} })
```
and there is still another one...

sachseka reviewed (changes requested):
I'm pretty sure (though not entirely) that using JKerr does not work. Take your Example 3: JKerr (equa.rasch.jk2—by the way, this is JK1, isn't it?!) throws a linking error for e12:
```
N.items N.units     shift        SD linkerror.jackknife SE.SD.jackknife
1      36       4 0.2430037 0.4934753           0.1155699        0.141694
```

However, in your output, there is still the usual linking error as given by equa.rasch:
```
$descriptives
  N.Items        SD        Var  linkerror
1      34 0.2456346 0.06033637 0.04212599
```

As you can see from your output:

```
lErrors2 <- multiEquatError(results2[[1]], results2[[2]], results2[[3]], difBound = 0.64)
Dimension 'domainlistening': Direct linking error of the three combinations of measurements: 
        mzp1 mzp2 N.Items    SD   Var linkerror chained mzp_1.vs.3 mzp_1.vs.2.vs.3
      1    1    2      34 0.246 0.060     0.042      NA         NA              NA
```
and also
```
jkErr2  <- multiEquatError(results2[[1]], results2[[2]], results2[[3]], difBound = 0.64, testletStr = tl.frame)
The 'unit' column in 'testletStr' seemed to be called 'testlet'. Rename this column into 'unit'.

Dimension 'domainlistening': Direct linking error of the three combinations of measurements: 
        mzp1 mzp2 N.Items    SD   Var linkerror chained mzp_1.vs.3 mzp_1.vs.2.vs.3
      1    1    2      34 0.246 0.060     0.042      NA         NA              NA
```

Again, I strongly recommend tests. What could be a plan to achieve this despite our tight time constraints?

weirichs commented:
Ok, I can add some tests. 

After looking at the function a little more closely, I would say that it is unfortunately written in a rather ambiguous and unnecessarily cumbersome way. I would strongly advocate fundamentally rewriting it so that it is much shorter and more transparent, but this would mean that the user input would also change. Specifically, I would not pass data.frames to the function with `x1`, `x2`, `x3`, but objects returned from `equat1pl`. They already have a defined structure. When `equat1pl` is called, the user must then already decide whether he wants to use testlets, i.e. determine whether to compute the linking errors by means of jackknife. `multiEquatError()` would then no longer have to deal with this. I would find this much more transparent. But as I said, this would then no longer be backwards-compatible. Can we perhaps talk about this briefly tomorrow? I have a bad feeling that it's not a good idea to just decide that now.

sachseka commented:
> Ok, I can add some tests.
> 
> After looking at the function a little more closely, I would say that it is unfortunately written in a rather ambiguous and unnecessarily cumbersome way. I would strongly advocate fundamentally rewriting it so that it is much shorter and more transparent, but this would mean that the user input would also change. Specifically, I would not pass data.frames to the function with `x1`, `x2`, `x3`, but objects returned from `equat1pl`. They already have a defined structure. When `equat1pl` is called, the user must then already decide whether he wants to use testlets, i.e. determine whether to compute the linking errors by means of jackknife. `multiEquatError()` would then no longer have to deal with this. I would find this much more transparent. But as I said, this would then no longer be backwards-compatible. Can we perhaps talk about this briefly tomorrow? I have a bad feeling that it's not a good idea to just decide that now.

Yes, I fully support discussing fundamental rewriting. However, I would prefer to retain the flexibility of passing data frames, as we do not always have an equat1pl object available. Would it be possible for the function to internally apply equat1pl to data frames automatically?

weirichs commented:
> Yes, I fully support discussing fundamental rewriting. However, I would prefer to retain the flexibility of passing data frames, as we do not always have an equat1pl object available. Would it be possible for the function to internally apply equat1pl to data frames automatically?

In theory yes, but that's exactly what made this function so unwieldy and confusing. The fact that it internally repeated everything that had already been done, i.e. these pairwise links, including the variant with or without testlets. It would be exactly the same as if runModels ran through the whole defineModel-preparation again. I would also find it more transparent if the function only did what it claims to do, namely a chained combination of already given linking errors, and not a recalculation and subsequent calculation. 

If an equal1pl object isn't available, it can easily be created by calling equat1pl, even if only two data.frames are available. equat1pl does not necessarily require the output of the getResults function.

weirichs committed:
multiEquatError rewritten

tests added, help files adapted

weirichs committed:
remove assertions fixSlopeMat

In the assertions, the columns must not be selected according to order, as this can vary.

sachseka reviewed (approved):
I didn't have the time to check it all but if you want to merge you can merge.

weirichs committed:
remove directory assertions
